pkg/generator/scheduler.go

```diff
@@ -0,0 +1,151 @@
+package generator
+
+import (
+	"context"
+	"sync"
+)
+
+type state int
+
+const (
+	working state = iota
+	stopped
+)
+
+// Protocol defines the interface that allows the Scheduler to determine if the
+// protocol is executing or not. This interface should be implemented by all
+// important protocols of the client, such as distributed key generation or
+// signing.
+type Protocol interface {
+	IsExecuting() bool
+}
+
+// Scheduler allows managing computationally heavy operations: stopping and
+// resuming them. The client needs to generate parameters for cryptographic
+// algorithms and generating these parameters requires a lot of CPU cycles.
+// The generation process may starve other processes in the client when it comes
+// to access to the CPU. The scheduler allows starting the parameter generation
+// when no other processes such as key generation or signing are executed on the
+// client. This way, the client that would normally be idle, can spend CPU
+// cycles on computationally heavy operations and stop these operations when CPU
+// cycles are needed elsewhere.
+type Scheduler struct {
+	state     state
+	workers   []func(context.Context)
+	stops     []context.CancelFunc
+	workMutex sync.Mutex
+
+	protocols      []Protocol
+	protocolsMutex sync.Mutex
+}
+
+func (s *Scheduler) RegisterProtocol(protocol Protocol) {
```

Comment:
Can we add a docstring for this public function?

Reply:
https://github.com/keep-network/keep-core/commit/91fc27a77c9976bf147aa942a2d08dd9940c267d